File: packages/poe-agent/src/runtime/transcript.ts
import path from "node:path";
import type { SessionUpdate } from "@poe-code/poe-acp-client";
import { hasOwnErrorCode } from "../error-codes.js";
import type { AcpEvent } from "./types.js";

export function mapAcpEventToSessionUpdates(event: AcpEvent): SessionUpdate[] {
  if (event.type === "message.delta") {
    if (event.content.length === 0) return [];
    return [{ sessionUpdate: "agent_message_chunk", content: { type: "text", text: event.content } }];
  }

  if (event.type === "tool.intent") {
    return [
      {
        sessionUpdate: "tool_call",
        toolCallId: event.intentId,
        title: event.tool,
        kind: "execute",
        status: "pending",
        rawInput: event.args,
      },
      {
        sessionUpdate: "tool_call_update",
        toolCallId: event.intentId,
        kind: "execute",
        status: "in_progress",
      },
    ];
  }

  if (event.type === "tool.result") {
    return [
      {
        sessionUpdate: "tool_call_update",
        toolCallId: event.intentId,
        kind: "execute",
        status: "completed",
        rawOutput: event.result,
      },
    ];
  }

  if (event.type === "tool.error") {
    return [
      {
        sessionUpdate: "tool_call_update",
        toolCallId: event.intentId,
        kind: "execute",
        status: "failed",
        rawOutput: event.error,
      },
    ];
  }

  if (event.type === "usage") {
    const { inputTokens, outputTokens, cachedTokens, cacheCreationTokens } = event.usage;
    const nonCachedInput = Math.max(0, inputTokens - cachedTokens);
    return [
      {
        sessionUpdate: "usage_update",
        used: nonCachedInput,
        size: inputTokens,
        _meta: { inputTokens, outputTokens, cachedTokens, cacheCreationTokens },
      },
    ];
  }

  return [];
}

export interface TranscriptWriter {
  write(event: AcpEvent): Promise<void>;
  close(): Promise<void>;
  readonly filePath: string;
}

export interface TranscriptFsApi {
  mkdir(dir: string, options: { recursive: true }): Promise<void>;
  appendFile(path: string, contents: string): Promise<void>;
  lstat(path: string): Promise<{ isSymbolicLink(): boolean }>;
}

export interface CreateTranscriptWriterOptions {
  logPath?: string;
  logDir?: string;
  logFileName?: string;
  fs: TranscriptFsApi;
  pathJoin?: (...parts: string[]) => string;
}

export function createTranscriptWriter(
  options: CreateTranscriptWriterOptions,
): TranscriptWriter {
  const join = options.pathJoin ?? path.join;
  const filePath = resolveTranscriptFilePath(options, join);
  let dirEnsured: Promise<void> | undefined;
  const logDir = path.dirname(filePath);

  const ensureDir = (): Promise<void> => {
    if (!dirEnsured) {
      dirEnsured = options.fs.mkdir(logDir, { recursive: true });
    }
    return dirEnsured;
  };

  return {
    filePath,
    async write(event: AcpEvent): Promise<void> {
      const updates = mapAcpEventToSessionUpdates(event);
      if (updates.length === 0) return;

      await ensureNoSymbolicLinkPath(options.fs, filePath);
      await ensureDir();
      await ensureNoSymbolicLinkPath(options.fs, filePath);
      const payload = updates.map(update => `${JSON.stringify(update)}\n`).join("");
      await options.fs.appendFile(filePath, payload);
    },
    async close(): Promise<void> {
      // No persistent handle: appendFile opens/closes each write. Nothing to do.
    },
  };
}

async function ensureNoSymbolicLinkPath(fs: TranscriptFsApi, filePath: string): Promise<void> {
  const absolutePath = path.resolve(filePath);
  const root = path.parse(absolutePath).root;
  let inspectedPath = root;

  for (const segment of absolutePath.slice(root.length).split(path.sep).filter(Boolean)) {
    inspectedPath = path.join(inspectedPath, segment);
    try {
      if ((await fs.lstat(inspectedPath)).isSymbolicLink()) {
        throw new Error(`Transcript log path may not contain symbolic links: ${filePath}`);
      }
    } catch (error) {
      if (hasOwnErrorCode(error, "ENOENT")) {
        return;
      }
      throw error;
    }
  }
}

function resolveTranscriptFilePath(
  options: CreateTranscriptWriterOptions,
  join: (...parts: string[]) => string,
): string {
  if (options.logPath) {
    return options.logPath;
  }

  if (options.logDir && options.logFileName) {
    return join(options.logDir, options.logFileName);
  }

  throw new Error("createTranscriptWriter requires logPath or logDir + logFileName.");
}

File: packages/poe-agent-rust/src/transcript.rs
//! ACP replay updates with opaque binding-host payload handles.
#[derive(Debug, PartialEq)]
pub enum Kind {
    Ignore,
    Message,
    Intent,
    Result,
    Error,
    Usage,
}
pub fn classify<F, E>(mut probe: F) -> Result<Kind, E>
where
    F: FnMut(&'static str) -> Result<bool, E>,
{
    for (label, kind) in [
        ("message.delta", Kind::Message),
        ("tool.intent", Kind::Intent),
        ("tool.result", Kind::Result),
        ("tool.error", Kind::Error),
        ("usage", Kind::Usage),
    ] {
        if probe(label)? {
            return Ok(kind);
        }
    }
    Ok(Kind::Ignore)
}
#[derive(Debug, PartialEq)]
pub enum Node<T> {
    Undefined,
    Bool(bool),
    String(&'static str),
    Utf16(Vec<u16>),
    Number(f64),
    Opaque(T),
    Object(Vec<(&'static str, Node<T>)>),
    Array(Vec<Node<T>>),
}
impl<T> Node<T> {
    pub fn try_map<U, E>(self, mapper: &mut impl FnMut(T) -> Result<U, E>) -> Result<Node<U>, E> {
        Ok(match self {
            Self::Undefined => Node::Undefined,
            Self::Bool(value) => Node::Bool(value),
            Self::String(value) => Node::String(value),
            Self::Utf16(value) => Node::Utf16(value),
            Self::Number(value) => Node::Number(value),
            Self::Opaque(value) => Node::Opaque(mapper(value)?),
            Self::Array(values) => Node::Array(
                values
                    .into_iter()
                    .map(|value| value.try_map(mapper))
                    .collect::<Result<_, _>>()?,
            ),
            Self::Object(fields) => Node::Object(
                fields
                    .into_iter()
                    .map(|(key, value)| Ok((key, value.try_map(mapper)?)))
                    .collect::<Result<_, E>>()?,
            ),
        })
    }
}
pub enum Event<T> {
    Ignore,
    Message {
        content: T,
        empty: bool,
    },
    Intent {
        id: T,
        subsequent_id: T,
        tool: T,
        args: T,
    },
    Result {
        id: T,
        result: T,
    },
    Error {
        id: T,
        error: T,
    },
    Usage {
        input: T,
        output: T,
        cached: T,
        creation: T,
        difference: f64,
    },
}
pub fn updates<T: Clone>(event: Event<T>) -> Vec<Node<T>> {
    use Node::{Number, Object, Opaque, String};
    match event {
        Event::Ignore | Event::Message { empty: true, .. } => vec![],
        Event::Message { content, .. } => vec![Object(vec![
            ("sessionUpdate", String("agent_message_chunk")),
            (
                "content",
                Object(vec![("type", String("text")), ("text", Opaque(content))]),
            ),
        ])],
        Event::Intent {
            id,
            subsequent_id,
            tool,
            args,
        } => vec![
            Object(vec![
                ("sessionUpdate", String("tool_call")),
                ("toolCallId", Opaque(id.clone())),
                ("title", Opaque(tool)),
                ("kind", String("execute")),
                ("status", String("pending")),
                ("rawInput", Opaque(args)),
            ]),
            Object(vec![
                ("sessionUpdate", String("tool_call_update")),
                ("toolCallId", Opaque(subsequent_id)),
                ("kind", String("execute")),
                ("status", String("in_progress")),
            ]),
        ],
        Event::Result { id, result } => vec![Object(vec![
            ("sessionUpdate", String("tool_call_update")),
            ("toolCallId", Opaque(id)),
            ("kind", String("execute")),
            ("status", String("completed")),
            ("rawOutput", Opaque(result)),
        ])],
        Event::Error { id, error } => vec![Object(vec![
            ("sessionUpdate", String("tool_call_update")),
            ("toolCallId", Opaque(id)),
            ("kind", String("execute")),
            ("status", String("failed")),
            ("rawOutput", Opaque(error)),
        ])],
        Event::Usage {
            input,
            output,
            cached,
            creation,
            difference,
        } => vec![Object(vec![
            ("sessionUpdate", String("usage_update")),
            (
                "used",
                Number(if difference.is_nan() {
                    f64::NAN
                } else {
                    if difference <= 0.0 { 0.0 } else { difference }
                }),
            ),
            ("size", Opaque(input.clone())),
            (
                "_meta",
                Object(vec![
                    ("inputTokens", Opaque(input)),
                    ("outputTokens", Opaque(output)),
                    ("cachedTokens", Opaque(cached)),
                    ("cacheCreationTokens", Opaque(creation)),
                ]),
            ),
        ])],
    }
}
